core/src/main/java/io/grpc/ClientCallImpl.java

```diff
@@ -127,49 +144,122 @@ public void start(Listener<RespT> observer, Metadata.Headers headers) {
       headers.put(HttpUtil.USER_AGENT_KEY, userAgent);
     }
 
+    boolean started = startInternal(listener, headers);
+    // Start the deadline timer after stream creation because it will close the stream
+    if (started && deadlineNanoTime != null) {
+      deadlineCancellationFuture = startDeadlineTimer(timeoutMicros);
+    }
+  }
+
+  private boolean startInternal(
+      final ClientStreamListener listener, final Metadata.Headers headers) {
+    ClientTransport transport;
     try {
-      stream = transport.newStream(method, headers, listener);
+      transport = clientTransportProvider.get();
+    } catch (RuntimeException ex) {
+      closeCallPrematurely(listener, Status.fromThrowable(ex));
+      return false;
+    }
+    if (transport == null) {
+      closeCallPrematurely(listener, Status.UNAVAILABLE.withDescription("Channel is shutdown"));
+      return false;
+    }
+
+    try {
+      synchronized (this) {
+        stream = transport.newStream(method, headers, listener);
+      }
+      // This will typically be empty, and no calls will be added to it once stream is not null.
+      for (Runnable queuedCall : queuedCalls) {
+        queuedCall.run();
+      }
     } catch (IllegalStateException ex) {
       // We can race with the transport and end up trying to use a terminated transport.
       // TODO(ejona86): Improve the API to remove the possibility of the race.
+      retryFuture = scheduledExecutor.schedule(new Runnable() {
+        @Override
+        public void run() {
+          startInternal(listener, headers);
+        }
+      }, (long) getNextDelayMillis(), TimeUnit.MILLISECONDS);
+
       closeCallPrematurely(listener, Status.fromThrowable(ex));
     }
-    // Start the deadline timer after stream creation because it will close the stream
-    if (deadlineNanoTime != null) {
-      deadlineCancellationFuture = startDeadlineTimer(timeoutMicros);
-    }
+
+    return true;
   }
 
   @Override
-  public void request(int numMessages) {
-    Preconditions.checkState(stream != null, "Not started");
+  public void request(final int numMessages) {
+    checkState(state == State.STARTED || state == State.HALF_CLOSED,
+        "Call was either not started or canceled");
+    synchronized (this) {
+      if (stream == null) {
+        queuedCalls.add(new Runnable() {
+          @Override
+          public void run() {
+            request(numMessages);
```

Comment:
Isn't the state-checking broken like that though?

```java
call.sendPayload(o);
call.halfClosed();
```

Let's say you have the above (valid) user code. If the transport is initially down, then by the time the transport is up the call is already in half-closed state. But when sendPayload's Runnable is executed, it will throw.